software/contrib/master_clock.py

```diff
@@ -0,0 +1,340 @@
+        await asyncio.sleep_ms(self.pulseWidthMs)
+        cv.off()
+
+    ''' Given a desired BPM, calculate the time to sleep between clock pulses '''
+    def calcSleepTime(self):
+        self.timeToSleepMs = int((60000 / self.bpm / self.CLOCKS_PER_QUARTER_NOTE))
+    
+    def checkForAinBPM(self):
+        val = 100 * ain.percent()
+        # If there is an analogue input voltage use that for BPM
+        if val > self.minAnalogInputVoltage:
+            self.bpm = int((((self.MAX_BPM) / 100) * val) + self.MIN_BPM)
+            self.calcSleepTime()
+            self.getPulseWidth()      
+
+    def getBPM(self):
+        self.bpm = self.MIN_BPM + k1.read_position(steps=(self.MAX_BPM - self.MIN_BPM + 1), samples=512)
+
+    def getPulseWidth(self):
+        self.MAX_PULSE_WIDTH = self.timeToSleepMs // 2  # Maximum of 50% pulse width
+        if self.configMode:
+            # Get desired PW percent
+            self.pulseWidthPercent = k2.read_position(steps=50, samples=512) + 1
+            # Calc Pulse Width duration (x 2 needed because the max is 50%)
+            self.pulseWidthMs = int((self.MAX_PULSE_WIDTH * 2) * (self.pulseWidthPercent / 100)) 
+            # Don't allow a pulse width less than the minimum
+            if self.pulseWidthMs < self.MIN_PULSE_WIDTH:
+                self.pulseWidthMs = self.MIN_PULSE_WIDTH
+                self.pulseWidthPercent = 'min'
+        else:
+            self.pulseWidthMs = int((self.MAX_PULSE_WIDTH * 2) * (self.pulseWidthPercent / 100))
+
+    ''' Triggered by main. Sends output pulses at required division '''
+    def clockTrigger(self):
+
+        if self.DEBUG:
+            print('BPM: ' + str(self.bpm) + ' cycle: ' + str(self.timeToSleepMs) + ' PW:' + str(self.pulseWidthMs))
+
+        self.el.create_task(self.outputPulse(cv1))
+
+        if self.divisionOutput2 != 0:
+            if self.step % self.divisionOutput2 == 0:
+                self.el.create_task(self.outputPulse(cv2))
+        else:
+            # 0 = random
+            cv2.value(randint(0, 1))
+
+        if self.divisionOutput3 != 0:
+            if self.step % self.divisionOutput3 == 0:
+                self.el.create_task(self.outputPulse(cv3))
+        else:
+            # 0 = random
+            cv3.value(randint(0, 1))
+
+        if self.divisionOutput4 != 0:
+            if self.step % self.divisionOutput4 == 0:
+                self.el.create_task(self.outputPulse(cv4))
+        else:
+            # 0 = random
+            cv4.value(randint(0, 1))
+
+        if self.divisionOutput5 != 0:
+            if self.step % self.divisionOutput5 == 0:
+                self.el.create_task(self.outputPulse(cv5))
+        else:
+            # 0 = random
+            cv5.value(randint(0, 1))
+
+        if self.divisionOutput6 != 0:
+            if self.step % self.divisionOutput6 == 0:
+                self.el.create_task(self.outputPulse(cv6))
+        else:
+            # 0 = random
+            cv6.value(randint(0, 1))
```

Comment:
[optional] The above section of code governing cv2-6 is just begging to be a for loop.

Reply:
I can't intuitively think of a for loop that would be more efficient or easier to read. I am open to suggestions though,

Reply:
I think it would require a bit more refactoring by putting the division outputs into a list instead of individual class attributes, then enumerating over them and [europi.cvs](https://github.com/Allen-Synthesis/EuroPi/blob/main/software/firmware/europi.py#L528):
```python
for idx, output in enumerate(self.divisionOutputs):
  if output and self.step % output == 0:
    self.el.create_task(self.outputPulse(cvs[idx])
  else:
    cvs[idx].value(randint(0, 1)
``` 

That would clean up other code sections like159-170

```python
# Only adjust values if k2 has moved. This avoids a potentially annoying UX
if self.previousselectedDivision != selectedDivision:
    self.divisionOutputs[self.activeOption] = selectedDivision
    self.saveState()

self.previousselectedDivision = selectedDivision
```

Looking over all the instances of `self.divisionOutputX`, I think it's doable, but would be a bit of a significant change to the script.

Reply:
Thanks. you were right the change was significant, but doing this streamlined a few other areas too! Done!